### ISSUE TITLE
Parse forge mods from status response

### ISSUE THREAD

**Issue by PerchunPak** [py-mine/mcstatus#556]
This includes `forgeData` key in the answer and other modded cores, if they provide this information.

**Comment by PerchunPak** [comment]
> `forgeData` isn't a part of minecraft's official protocol, hence mcstatus doesn't support it. However `forgeData` seems to just be a part of the status response, as just an additional key. This means that it's already being obtained and is accessible, why would having a bool in the reader/writer classes help?
>
> Mcstatus probably isn't the place to introduce these kinds of changes, as you're not expected to use these classes in your code. These classes are considered internal, and they are not a part of the public API. This means they're only expected to be used for internal functionalities, and if they're not necessary for those, they're not needed.
>
> If you do need this kind of change, either implement it locally within your project, or consider checking out https://github.com/py-mine/mcproto, which already supports this behavior, and where the reader/writer classes are indeed a part of the public API: see the implementation ([enum def](https://github.com/py-mine/mcproto/blob/f2874148c18102efcf8746014749e77e0056b4ca/mcproto/protocol/base_io.py#L38), [writer](https://github.com/py-mine/mcproto/blob/f2874148c18102efcf8746014749e77e0056b4ca/mcproto/protocol/base_io.py#L104) and [reader](https://github.com/py-mine/mcproto/blob/f2874148c18102efcf8746014749e77e0056b4ca/mcproto/protocol/base_io.py#L344)).

_Originally posted by @ItsDrike in https://github.com/py-mine/mcstatus/issues/555#issuecomment-1565999619_

**Comment by PerchunPak** [comment]
> `forgeData` isn't a part of minecraft's official protocol, hence mcstatus doesn't support it.

We support modded server cores as well as vanilla one (#446, #433, etc). Although, they usually just follow the protocol, we sometimes change some things to fit their needs (#439, we didn't have a required part of request, but vanilla was fine with it, so why we would change it?).

> However `forgeData` seems to just be a part of the status response, as just an additional key.

It's and it doesn't have documentation which makes it hard to general user to understand it (it has quite strange structure and multiple fully different versions).

> These classes are considered internal, and they are not a part of the public API.

They are, but they are also often used and referenced. Both mcstatus.io and mcsrvstat.us parse `forgeData` property and show this information to user, so would we don't do this?

**Comment by ItsDrike** [comment]
> They are, but they are also often used and referenced. Both mcstatus.io and mcsrvstat.us parse forgeData property and show this information to user, so would we don't do this?}

Perhaps, however as `forgeData` is just a part of status, I still don't understand why a bool reader/writer is necessary. It's not like we need to read an additional packet that contains this data, we already read it as a part of status, it's essentially just another JSON object. Does it contain some bytearrays that need to be read with a buffer (or as we call it connection) reader? Or what's the reason we would need this additional bool operations?

**Comment by ItsDrike** [comment]
> It's and it doesn't have documentation which makes it hard to general user to understand it (it has quite strange structure and multiple fully different versions).

This sounds concerning, I don't necessarily love the idea of adding support for parsing something that might constantly end up changing with newer versions in major ways. But it depends on what these changes are, how many of them we'd need to support at a time, and other factors.

> They are, but they are also often used and referenced. Both mcstatus.io and mcsrvstat.us parse forgeData property and show this information to user, so would we don't do this?

In this comment, I was talking about the `Connection` classes, and the bool reader/writers. These are fully internal and nobody should be using them. `forgeData` should be available as a part of `raw` of the status response class. This is actually a part of the public API already.

**Comment by CoolCat467** [comment]
I've actually already written a forgeData response reader for my discord bot StatusBot [linked here](https://github.com/CoolCat467/StatusBot/blob/main/src/StatusBot/decode_mods.py), it would just be a matter of making a nice interface for it here

**Comment by PerchunPak** [comment]
Why do you need to transform dict object to buffer and use bytes? Why can't you just use dict?

**Comment by PerchunPak** [comment]
Actually, it has quite poor [documentation](https://wiki.vg/Minecraft_Forge_Handshake), but we anyway need to do our own research.

**Comment by ItsDrike** [comment]
> Why do you need to transform dict object to buffer and use bytes? Why can't you just use dict?

From the example code CoolCat467 sent, it seems that `forgeData` dict (originally received as JSON), contains some data under the "d" key, which while sent as string in the JSON, is actually interpreted as binary data, and you do indeed need the buffer classes to properly read it out (See [this line](https://github.com/CoolCat467/StatusBot/blob/34ffced1238f6a8ac754a1d1641f1544b90b5c90/src/StatusBot/decode_mods.py#L86) in the linked file).

**Comment by PerchunPak** [comment]
> `forgeData` dict (originally received as JSON), contains some data under the "d" key

I tested it with `grmpixelmon.com` (Forge 1.16.5, it's a just some public server that I use to test mcstatus sometimes) and didn't get any `d` key anywhere. I'm confused where it comes from and what it means.

**Comment by ItsDrike** [comment]
Yeah, I also didn't find any good documentation for it, do you have some docs or even source code link to the forge implementation that describes this @CoolCat467?

**Comment by CoolCat467** [comment]
Yes, let me try to go find it. I was meaning to find the link to this the other day but it was getting a bit late.

**Comment by CoolCat467** [comment]
Here we go, this file from forge explains everything pretty well:
https://github.com/MinecraftForge/MinecraftForge/blob/42115d37d6a46856e3dc914b54a1ce6d33b9872a/src/main/java/net/minecraftforge/network/ServerStatusPing.java

**Comment by PerchunPak** [comment]
Oh, that makes everything much more cleaner... I will approve #555 after you will resolve the conversation.

**Comment by ItsDrike** [comment]
This issue has the WIP label, but I no progress was done on it for 2 years, I think it's fair to say this isn't a work-in-progress at this point, unless someone wants to pick it up.

**Comment by CoolCat467** [comment]
If you would like I could make a pull request adding the previously linked decoder module I've made, but I'm not promising to do anything beyond that.

**Comment by ItsDrike** [comment]
> If you would like I could make a pull request adding the previously linked decoder module I've made, but I'm not promising to do anything beyond that.

If it works consistently across all forge versions, then sure, that'd be great!

**Comment by CoolCat467** [comment]
Oh wait it looks like that got handled as a part of https://github.com/py-mine/mcstatus/pull/578

**Comment by ItsDrike** [comment]
oh lol, I thought it was somewhere, but since nothing was linked, I figured I just imagined it, guess not, in that case, this can probably just be closed.
